pinch-zoom-canvas.js

```diff
@@ -1,468 +1,528 @@
+
+      // draws the image position and scale
+      this.context.drawImage(
+        this.imgTexture,
+        this.position.x, this.position.y,
+        this.scale.x * this.imgTexture.width,
+        this.scale.y * this.imgTexture.height)
+
+      requestAnimationFrame(this.render)
+    },
+
+    pause: function () {
+      this.running = false
+      return this
+    },
+
+    resume: function () {
+      this.calculateOffset()
+      this.running = true
+      requestAnimationFrame(this.render)
+      return this
+    },
+
+		/**
+		 * Calculates the offset of the canvas position relative to the page
+     * since touch events are always relative to the page
+		 */
+    calculateOffset: function () {
+      if (!this.canvas) return this
+
+      // no offset if we're using direct client coordinates (full screen canvas)
+      if (!this.fullScreen) {
+        const canvasBox = this.canvas.getBoundingClientRect()
+
+        // caculate the offset from the scroll position
+        this.offset.x = canvasBox.left + window.scrollX
+        this.offset.y = canvasBox.top + window.scrollY
+      }
+
+      return this
+    },
+
+    /**
+     * handles zooming in and out
+     */
+    zoom: function (zoom, touchX, touchY) {
+      if (!zoom) return
+
+      //new scale
+      var currentScale = this.scale.x
+      var newScale = this.scale.x + zoom / 100
+
+      // TODO: Make bounceback on zoomed to in or out instead of hard setting
+      if (newScale < this.initialScale) { // we are below the minimum zoom (initialZoom)
+        this.zoomed = false // we're back at the initial scale
+        // this.position.x = this.initPosition.x
+        // this.position.y = this.initPosition.y
+        // this.scale.x = this.initialScale
+        // this.scale.y = this.initialScale
+        newScale = this.initialScale // set to the initialScale
+      } else if (this.maxZoom && newScale > this.maxZoom) { // we are above maximum zoom
+        this.zoomed = true
+        newScale = this.maxZoom // set to maximum zoom
+      } else { // we are zoomed in between min and max
+        this.zoomed = true
+      }
+
+      var deltaScale = newScale - currentScale
+      var currentWidth = this.imgTexture.width * this.scale.x
+      var currentHeight = this.imgTexture.height * this.scale.y
+      var deltaWidth = this.imgTexture.width * deltaScale
+      var deltaHeight = this.imgTexture.height * deltaScale
+
+      var tX = (touchX * 2 - this.position.x)
+      var tY = (touchY * 2 - this.position.y)
+      var pX = -tX / currentWidth
+      var pY = -tY / currentHeight
+
+
+      //finally affectations
+      this.scale.x = newScale
+      this.scale.y = newScale
+      this.position.x += pX * deltaWidth
+      this.position.y += pY * deltaHeight
+
+      // onZoom callback
+      if (this.onZoom) {
+        this.onZoom(newScale, this.zoomed)
+      }
+
+    },
+
+    move: function (relativeX, relativeY) {
```

Comment:
this should remain public as you can call it externally to force move things outside of user input